transport/session/src/balancer/rate_limiting.rs

```diff
@@ -0,0 +1,478 @@
+use std::future::Future;
+use std::pin::Pin;
+use std::sync::atomic::{AtomicU64, Ordering};
+use std::sync::Arc;
+use std::task::{Context, Poll};
+use std::time::{Duration, Instant};
+
+use futures::Stream;
+use futures_time::task::Sleep;
+use pin_project::pin_project;
+
+/// Controller for [`RateLimitedStream`] to allow dynamic controlling of the stream's rate.
+#[derive(Debug)]
+pub struct RateController(Arc<AtomicU64>);
+
+#[allow(unused)]
+impl RateController {
+    const MIN_DELAY: Duration = Duration::from_micros(100);
+
+    /// Update the rate limit (elements per unit).
+    pub fn set_rate_per_unit(&self, elements_per_unit: usize, unit: Duration) {
+        assert!(unit > Duration::ZERO, "unit must be greater than zero");
+
+        if elements_per_unit > 0 {
+            // Calculate the next allowable time based on the rate
+            let rate_per_sec = (elements_per_unit as f64 / unit.as_secs_f64());
+
+            // Convert to duration (seconds per element) and subtract the duration
+            // it took to yield the element from the inner stream
+            let new_rate = Duration::from_secs_f64(1.0 / rate_per_sec)
+                .max(Self::MIN_DELAY)
+                .as_micros()
+                .min(u64::MAX as u128) as u64; // Clamp to u64 to avoid overflow
+
+            self.0.store(new_rate, Ordering::Relaxed);
+        } else {
+            self.0.store(0, Ordering::Relaxed);
+        }
+    }
+
+    /// Get the current rate limit per time unit.
+    pub fn get_rate_per_sec(&self) -> f64 {
+        let delim = self.0.load(Ordering::Relaxed);
+        if delim > 0 {
+            1.0 / Duration::from_micros(delim).as_secs_f64()
+        } else {
+            0.0
+        }
+    }
+}
+
+enum State {
+    Read,
+    NoRate,
+    Wait,
+}
+
+/// A stream adapter that yields elements at a controlled rate, with dynamic rate adjustment.
+///
+/// See [`RateLimitExt::rate_limit_per_unit`].
+#[pin_project]
+pub struct RateLimitedStream<S: Stream> {
```

Comment:
Does something like this not exist yet?

Reply:
Unfortunately I didn't find any, that would have runtime adjustable rate control. Their rate-limit can be set only once on construction.